src/components/vapp.ts

```diff
@@ -0,0 +1,263 @@
+import * as paper from 'paper';
+import { EntityLabelComponent } from './entity-label';
+import { VmData } from './vm';
+import { VmAndVnicListComponent } from './vm-and-vnic-list';
+import { VAPP_BACKGROUND_COLOR } from '../constants/colors';
+import { CONNECTOR_RADIUS, VIEW_BOTTOM_PADDING, DEFAULT_SCROLLBAR_THICKNESS, VAPP_NETWORK_RIGHT_MARGIN, VAPP_PADDING,
+  DEFAULT_STROKE_WIDTH } from '../constants/dimensions';
+import { MarginComponent } from './margin';
+import { VappNetworkData } from './vapp-network';
+import { VappNetworkListComponent } from './vapp-network-list';
+import { ScrollbarComponent } from './scrollbar';
+
+const MARGIN_RIGHT = 30;
+const BACKGROUND_RADIUS = 5;
+const LABEL_ICON_COLOR = '#CA67B8';
+const VAPP_LABEL_BOTTOM_MARGIN = 10;
+const EDGE_LABEL_BOTTOM_MARGIN = 5;
+
+/**
+ * Interface for vApp data.
+ */
+export interface VappData {
+  uuid: string;
+  name: string;
+  vapp_networks: VappNetworkData[];
+  vms: VmData[];
+}
+
+/**
+ * Virtual Application Visual Component.
+ */
+export class VappComponent extends paper.Group {
+
+  readonly label: EntityLabelComponent;
+  readonly background: paper.Path.Rectangle;
+  readonly _margin: MarginComponent;
+  readonly vms: VmAndVnicListComponent;
+  readonly vappNetworks: VappNetworkListComponent;
+  // position for the division between any labels and vm/vnic list
+  readonly divisionPositionY: number;
+  private scrollbar: ScrollbarComponent;
+  // content needs scrollbar
+  private _isScrollable: boolean = false;
+
+  /**
+   * Creates a new VappComponent instance.
+   *
+   * @param _vapp the vapp data
+   * @param _point the location that the vm should be rendered at
+   */
+  constructor(private _vapp: VappData,
+              private _point: paper.Point = new paper.Point(0, 0)) {
+    super();
+    this.applyMatrix = false;
+    this.pivot = new paper.Point(0, 0);
+    this.position = _point;
+
+    let backgroundOffsetX = 0;
+    let backgroundOffsetY = 0;
+    const vappNetworkCount = this._vapp.vapp_networks.length;
+    const vmCount = this._vapp.vms.length;
+
+    // vapp label
+    // x position based on if there are any vApp networks (and how many) or vms
+    const labelPositionX = vappNetworkCount || vmCount
+      ? Math.max(1, vappNetworkCount) * VAPP_NETWORK_RIGHT_MARGIN + VAPP_PADDING
+      : VAPP_PADDING;
+    this.label = new EntityLabelComponent(
+      this._vapp.name,
+      LABEL_ICON_COLOR,
+      new paper.Point(labelPositionX, VAPP_PADDING),
+      'bold'
+    );
+    this.addChild(this.label);
+
+    // start vapp network paths - add a point to each network to have x position data for matching vnics
+    // and edge labels to have y position data for vapp header (which includes the vapp label and edge labels)
+    if (vappNetworkCount) {
+      this.vappNetworks = new VappNetworkListComponent(
+        this._vapp.vapp_networks,
+        new paper.Point(VAPP_PADDING + CONNECTOR_RADIUS, 0));
+      this.addChild(this.vappNetworks);
+    }
```

Comment:
max (furthest right) vapp network position can be derived from this component's bounds to pass to the vmAndVnicListComponent

Reply:
nevermind, edge labels can throw the width off

Reply:
simplifying even more by using the value straight from the vappNetworkList.lastNetworkPosition